client/sql: fix concurrent Txn batches in the face of retry errors

nvb opened the pull request:
Fixes #17197.

Before this change, retryable errors did not interact well with concurrent
transaction batches. This could lead to strange errors, deadlocks, or in
especially bad cases, unintended data being committed. This poor handling of
retryable errors made parallel statements using RETURNING NOTHING pretty
dangerous to use.

This change addresses these complications of concurrent batches. It begins by
adding extra logic to the `sessions` handling of the `parallelizeQueue`, bumping
a transaction's epoch in the case of retryable errors to assure that no
unintended data is ever committed. This has the effect of fixing parallel
statements in the face of all retryable errors except for ABORT errors.

It turns out that abort errors were even more tricky because of the interaction
between `client.Txn` and `TxnCoordSender`. Previously, in cases where abort
errors were seen, the `TxnCoordSender` would stop heartbeating the transaction
and forget about it. However, if a successful response or even a non-abort retry
error for the old transaction ID was found later, the `TxnCoordSender` would
begin heartbeating again, believing that this was seeing a new Transaction. This
could quickly result in a deadlock if the next incarnation of the Txn began
waiting on an intent from the previous one. The root problem here is that
`TxnCoordSender` has no memory, so it's up to the `client.Txn` to prevent this.
In doing so, things are a little tricky because the response handling under lock
is split between `TxnCoordSender` and `client.Txn`, so responses can be
reordered over the API boundary. This serves as a further example of why
`client.Txn` and `TxnCoordSender` should be merged.

After fixing these issues, this change adds a few fairly comprehensive tests
that stress RETURNING NOTHING in the face of retryable errors.

nvb committed:
sql: keep track of all parallel batch errors in ParallelizeQueue

Previously we only held onto the first error seen in a parallel batch.
Because there's no way to guarantee that the first error seen contains
the root cause of the error, we now hold onto and report all errors for
each parallel batch.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17627)
<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but probably you'll want a second pair of eyeballs.

---

Reviewed 4 of 4 files at r1, 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/sender.go, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVtVIHfsjmMx9s2NGI:-KrVtVIHfsjmMx9s2NGJ:b2cqgmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L42)):*
> ```Go
> 	// ForgetTxn assures that the transaction is no longer being tracked
> 	// by the TxnCoordSender. Aborting the transaction and stopping its
> 	// heartbeat loop if it is. Implementations should be idepotent, and
> ```

s/idep/idemp/

---

*[pkg/internal/client/txn.go, line 1071 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVtnqo1eqzuouQgzSH:-KrVtnqo1eqzuouQgzSI:b-s90w8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1071)):*
> ```Go
> 	_ = txn.updateStateOnRetryableErrLocked(
> 		ctx, *newErr,
> 		// We're passing the current ID as the "request"'s. In doing so, we're
> ```

This comment isn't new, but could you explain what's going on here?

---

*[pkg/internal/client/txn.go, line 1124 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVu8ZuDyHG0G5WXIaY:-KrVu8ZvTI4C-NBIiraj:b-40k1v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1124)):*
> ```Go
> 		// incarnation of this Txn. This should be handled by the TxnCoordSender
> 		// itself when it sees the abort error, but because responses can
> 		// be reordered between the TxnCoordSender and Txn.Send, it may
> ```

So what happened here was that `TxnCoordSender` drops state about a txn, but an outstanding request comes back, and then a new heartbeat loop is started? Yikesies.

---

*[pkg/sql/executor.go, line 1421 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVvdI_PykP2YFD20jg:-KrVvdI_PykP2YFD20jh:btdgiq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/executor.go#L1421)):*
> ```Go
> 			// fixed.
> 			//
> 			// TODO DURING REVIEW: Should we remove the restriction but synchronize
> ```

That *sounds* like the right thing to do, right?

---

*[pkg/sql/session.go, line 665 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVvuTFoNkUeEYOUHhK:-KrVvuTFoNkUeEYOUHhL:byxuo39) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/session.go#L665)):*
> ```Go
> 		// the transaction epoch to invalidate any writes performed by any workers
> 		// after the retry updated the txn's proto but before we synchronized. Note
> 		// that we don't Txn need lock the client.Txn because we're synchronized.
> ```

this sentence is off.

---

*[pkg/sql/txn_restart_test.go, line 135 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVw2ziDRz-TpAlSklP:-KrVw2ziDRz-TpAlSklQ:b-bukkww) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L135)):*
> ```Go
> 			}},
> 		}
> 		shuffle.Shuffle(injections)
> ```

But if it fails, you don't even know the order. Should have a seed and print it.

---

*[pkg/sql/txn_restart_test.go, line 605 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVwGl26ZK8PMAUGjWP:-KrVwGl26ZK8PMAUGjWQ:bbmtft0) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L605)):*
> ```Go
> // with parallel statements. We test that when a statement running in parallel with
> // others gets a retryable error, the other statements synchronize and retry without
> // leaving around any unintended data. This was the case before #17197 was addressed.
> ```

"leaving around" doesn't seem so bad, but it they were leaving around *committable* data -- might be worth adding that word.

---

*[pkg/sql/txn_restart_test.go, line 684 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVxPUjcAbHMn4gqOHh:-KrVxPUjcAbHMn4gqOHi:b-8k7h8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L684)):*
> ```Go
> 
> 					// Verify that the txns succeeded by reading the rows from each table.
> 					// Each should have written exactly one row.
> ```

How quickly does this fail without your fixes in this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-:-KrVxYz-KyZVKZusZYKT:bp1nxxo)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 4 files at r1, 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/sender.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWOEqUVp3ZUrCmkeen:-KrWOEqUVp3ZUrCmkeeo:b3kt6n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L33)):*
> ```Go
> // the Txn client and DistSQL hacky powers when handling errors that
> // happened on remote nodes or with concurrent requests.
> type SenderWithTxnCoordBackdoor interface {
> ```

The `Sender` abstraction is pretty well broken at this point: it's no longer safe to interpose another `Sender` between `client.Txn` and `TxnCoordSender` as the interface would imply. File a cleanup issue to make `client.Txn` take a `*TxnCoordSender` instead of a `Sender` (unless this all becomes moot first when Txn and TxnCoordSender are merged). 

---

*[pkg/sql/txn_restart_test.go, line 135 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVw2ziDRz-TpAlSklP:-KrWS-z_w66xHToLZW8V:b-j63an2) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L135)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

But if it fails, you don't even know the order. Should have a seed and print it.
</blockquote></details>

This comes for free if you add our randutil.SeedForTests() in main_test.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-:-KrWSlX3yk18mdv-car7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/txn_restart_test.go, line 135 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVw2ziDRz-TpAlSklP:-KrWTouHVBROxsB7Sqw5:b-focq05) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L135)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This comes for free if you add our randutil.SeedForTests() in main_test.go
</blockquote></details>

Not really, unless everything that happens before this test is deterministic (at least in its usage of the rand).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/txn_restart_test.go, line 135 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVw2ziDRz-TpAlSklP:-KrWck-7qds0twnbeHTg:bik5gm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L135)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not really, unless everything that happens before this test is deterministic (at least in its usage of the rand).
</blockquote></details>

Or if you're running this test in isolation. So it's not too useful for CI runs, but you could still use it for stress runs of this test. But anyway, being explicit in tests that deliberately use randomness is a good idea. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 4 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/sender.go, line 44 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWjekD4oSLPETs-hI_:-KrWjekD4oSLPETs-hIa:bk4xawz) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L44)):*
> ```Go
> 	// heartbeat loop if it is. Implementations should be idepotent, and
> 	// should be a no-op if the transaction is not being tracked.
> 	ForgetTxn(ctx context.Context, txn roachpb.Transaction)
> ```

if a txnID is sufficient here instead of the whole proto, I think it'd be cleaner to just take that.
If we stick we the proto, we're passing it my pointer everywhere else, so I think we should be consistent here too.

---

*[pkg/internal/client/txn.go, line 1106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWiNGm94FdRr7WAwON:-KrWiNGm94FdRr7WAwOO:b-1riw28) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1106)):*
> ```Go
> 		// didn't start heartbeating again because of this delayed error.
> 		txn.forgetTxnOnAbort(ctx, newTxn)
> 		return roachpb.NewError(&roachpb.RespForPrevTxnIDError{})
> ```

I'm not sure why we're returning this error here?
We don't return an error for the other cases (error for previous incarnations) - i.e. the wrong epoch. Why not swallow this?

I think returning the error is OK in the other place (the non-error case in txn.Send), but perhaps `TxnAbortedAsyncErr` would be a better name. And I'd suggest we try to unify this error with what the `TxnCoordSender` returns from `maybeRejectClientLocked()`. Does that make sense?

---

*[pkg/internal/client/txn.go, line 1137 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWfna7TtooCPz1ru_U:-KrWfna8bro4w1OZFYqB:b-gpr8eo) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1137)):*
> ```Go
> 		// TxnCoordSender.
> 		txn.mu.Proto = *newTxn
> 	} else /* restartErr */ {
> ```

`restartErr` is not a thing. I'd remove that inline comment.

---

*[pkg/internal/client/txn.go, line 1152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWfHGTVQPTpWOs4KWK:-KrWfHGTVQPTpWOs4KWL:b-sdx927) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1152)):*
> ```Go
> }
> 
> // forgetTxnOnAbort assures that is the Txn's wrapped sender is a TxnCoordSender,
> ```

s/is/if

---

*[pkg/roachpb/data.go, line 862 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWlWDEN3CxWVfF8pbs:-KrWlWDF7kG3DtGI1T7X:b95ny1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/roachpb/data.go#L862)):*
> ```Go
> }
> 
> // BumpEpoch increments the transaction's epoch, allowing for an in-place
> ```

I don't know what an "in-place restart" is :)
I'd say that it allows a client to discard all the previous writes before restarting and it is to be used in cases where those previous writes might have been performed at the wrong epoch.

---

*[pkg/sql/executor.go, line 1421 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVvdI_PykP2YFD20jg:-KrWl93nWkSqGl5NMT_0:bnlidbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/executor.go#L1421)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

That *sounds* like the right thing to do, right?
</blockquote></details>

Yes, I think that's what we should do, at least morally. But actually, was I confused when I wrote this? Can the `parallelizeQueue.Empty()` ever be `false` here, given that we're executing another statement that's not in the queue?

Anyway, if we add the synchronization here, I'd also leave a TODO about figuring out how to cancel those statements.
And also, I guess we should look at the error from `queue.synchronize()` and bail if it's not retryable. But then again maybe the whole thing doesn't make sense since the current contract for returning errors would imply that the queue is empty...

---

*[pkg/sql/parallel_stmts.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWcdmgW9M3aXEixVfX:-KrWcdmgW9M3aXEixVfY:b-g5jexe) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cdb18705a080ce916e01dc13d67f82b2922f98/pkg/sql/parallel_stmts.go#L44)):*
> ```Go
> //    executed before the plan added second.
> // 3. No plans will begin execution once an error has been seen until Wait is
> //    called to drain the plans and reset the error set state.
> ```

nit: s/error set/error-set (otherwise parsing is hard)

---

*[pkg/sql/parallel_stmts.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWdhP6-Y3Qi1Tj65nZ:-KrWdhP6-Y3Qi1Tj65n_:bvvdip1) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cdb18705a080ce916e01dc13d67f82b2922f98/pkg/sql/parallel_stmts.go#L56)):*
> ```Go
> 
> 	// errs is the set of error seen since the last call to ParallelizeQueue.Wait.
> 	// Referred to as the current "parallel batch's error set".
> ```

consider giving some hints here about how this set is going to be used, or what the higher level might want to do with it

---

*[pkg/sql/session.go, line 664 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWhqDNylEkPDBN_-6p:-KrWhqDNylEkPDBN_-6q:bco30we) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/session.go#L664)):*
> ```Go
> 		// incarnation of this transaction. Before doing so though, we need to bump
> 		// the transaction epoch to invalidate any writes performed by any workers
> 		// after the retry updated the txn's proto but before we synchronized. Note
> ```

add "(some of these writes might have been performed at the wrong epoch)".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-:-KrWntFv2aqRLb8FnOTW:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Thanks for the reviews

---

Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/sender.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWOEqUVp3ZUrCmkeen:-KratYA-JYsLzDrqtpqS:b-anhx7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L33)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The `Sender` abstraction is pretty well broken at this point: it's no longer safe to interpose another `Sender` between `client.Txn` and `TxnCoordSender` as the interface would imply. File a cleanup issue to make `client.Txn` take a `*TxnCoordSender` instead of a `Sender` (unless this all becomes moot first when Txn and TxnCoordSender are merged). 
</blockquote></details>

Done. #17657.

---

*[pkg/internal/client/sender.go, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVtVIHfsjmMx9s2NGI:-KrXAUKIelHQyHxXvRQi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L42)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

s/idep/idemp/
</blockquote></details>

Done.

---

*[pkg/internal/client/sender.go, line 44 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWjekD4oSLPETs-hI_:-KrXkwUg1uAoUdx-lxfp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L44)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if a txnID is sufficient here instead of the whole proto, I think it'd be cleaner to just take that.
If we stick we the proto, we're passing it my pointer everywhere else, so I think we should be consistent here too.
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 1071 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVtnqo1eqzuouQgzSH:-KrXlKane_WlQKJftFN2:b-czm2xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1071)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This comment isn't new, but could you explain what's going on here?
</blockquote></details>

It looks like DistSQL doesn't support concurrent transaction batches because it can't handle one of them getting an Abort error.

---

*[pkg/internal/client/txn.go, line 1106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWiNGm94FdRr7WAwON:-KrarffVdLPZEAfqqYgJ:bjotg3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1106)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm not sure why we're returning this error here?
We don't return an error for the other cases (error for previous incarnations) - i.e. the wrong epoch. Why not swallow this?

I think returning the error is OK in the other place (the non-error case in txn.Send), but perhaps `TxnAbortedAsyncErr` would be a better name. And I'd suggest we try to unify this error with what the `TxnCoordSender` returns from `maybeRejectClientLocked()`. Does that make sense?
</blockquote></details>

You're right, swallowing the error is the right thing to do. Also changed the name to `TxnAbortedAsyncErr`. I'm not as convinced about merging the errors, as the error returned from `maybeRejectClientLocked` can be a symptom of other issues as well.

---

*[pkg/internal/client/txn.go, line 1137 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWfna7TtooCPz1ru_U:-KrXAhEqPpMZH-P5tO3U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1137)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`restartErr` is not a thing. I'd remove that inline comment.
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 1152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWfHGTVQPTpWOs4KWK:-KrXAkCQzA-LQFDUllZK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/txn.go#L1152)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/is/if
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1421 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVvdI_PykP2YFD20jg:-KrauzgEwVb-Ero5usZ_:bw4z66) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/executor.go#L1421)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Yes, I think that's what we should do, at least morally. But actually, was I confused when I wrote this? Can the `parallelizeQueue.Empty()` ever be `false` here, given that we're executing another statement that's not in the queue?

Anyway, if we add the synchronization here, I'd also leave a TODO about figuring out how to cancel those statements.
And also, I guess we should look at the error from `queue.synchronize()` and bail if it's not retryable. But then again maybe the whole thing doesn't make sense since the current contract for returning errors would imply that the queue is empty...
</blockquote></details>

Let me know if my comment makes sense. 

Also, I'm getting rid of `parallelizeQueue.Empty()` because it's misleading. Statements could have already been run and their errors could be waiting for a synchronization to be cleared. If we went with the current approach, those errors could be left over for the next iteration of statements.

---

*[pkg/sql/parallel_stmts.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWcdmgW9M3aXEixVfX:-KrXB8ZrtOdLQdTw_QHe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cdb18705a080ce916e01dc13d67f82b2922f98/pkg/sql/parallel_stmts.go#L44)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/error set/error-set (otherwise parsing is hard)
</blockquote></details>

Done.

---

*[pkg/sql/parallel_stmts.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWdhP6-Y3Qi1Tj65nZ:-KrXBfLJQQdR14lYPzYp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cdb18705a080ce916e01dc13d67f82b2922f98/pkg/sql/parallel_stmts.go#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider giving some hints here about how this set is going to be used, or what the higher level might want to do with it
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 664 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWhqDNylEkPDBN_-6p:-KrXmrRoSxjRkK-1aDc8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/session.go#L664)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

add "(some of these writes might have been performed at the wrong epoch)".
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 665 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVvuTFoNkUeEYOUHhK:-KrXmsV-ZwwycNIr3pRf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/session.go#L665)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

this sentence is off.
</blockquote></details>

Done.

---

*[pkg/sql/txn_restart_test.go, line 135 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVw2ziDRz-TpAlSklP:-KrXnViS-8maDFFtVQj2:bmvrk1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L135)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Or if you're running this test in isolation. So it's not too useful for CI runs, but you could still use it for stress runs of this test. But anyway, being explicit in tests that deliberately use randomness is a good idea. 
</blockquote></details>

So you want me to call `randutil.SeedForTests()` again at the beginning of this test? It is already in sql's `TestMain`.

---

*[pkg/sql/txn_restart_test.go, line 605 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVwGl26ZK8PMAUGjWP:-KrXCXltDEJXBCsSHB4I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L605)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

"leaving around" doesn't seem so bad, but it they were leaving around *committable* data -- might be worth adding that word.
</blockquote></details>

Done.

---

*[pkg/sql/txn_restart_test.go, line 684 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVxPUjcAbHMn4gqOHh:-KrXC32brQwnjnlFWl5o:b-4qak57) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L684)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

How quickly does this fail without your fixes in this PR?
</blockquote></details>

Within 100 iterations, though we are injecting a lot of retry errors in each test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-:-KraxghxkCYXoKXUvPp9:b-9qdowt)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 13 of 13 files at r3, 13 of 13 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/txn_restart_test.go, line 135 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrVw2ziDRz-TpAlSklP:-Krb1QER1aAqMWNhAnyf:b-48ijs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/sql/txn_restart_test.go#L135)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

So you want me to call `randutil.SeedForTests()` again at the beginning of this test? It is already in sql's `TestMain`.
</blockquote></details>

Yeah, I guess (ideally `Shuffle` would accept an RNG). But I don't feel strongly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627)*
<!-- Sent from Reviewable.io -->

nvb committed:
testing: ReplayProtectionFilterWrapper shouldn't block on non-replays

nvb committed:
client/sql: fix concurrent Txn batches in the face of retry errors

Fixes #17197.

Before this change, retryable errors did not interact well with concurrent
transaction batches. This could lead to strange errors, deadlocks, or in
especially bad cases, unintended data being committed. This poor handling of
retryable errors made parallel statements using RETURNING NOTHING pretty
dangerous to use.

This change addresses these complications of concurrent batches. It does this by
adding extra logic to the `sessions` handling of the `parallelizeQueue`, bumping
a transaction's epoch in the case of retryable errors to assure that no
unintended data is ever committed. This has the effect of fixing parallel
statements in the face of all retryable errors except for ABORT errors.

After fixing this issues, this change adds a few fairly comprehensive tests
that stress RETURNING NOTHING in the face of retryable errors.

nvanbenschoten commented:
I realized that the extra hooks into `TxnCoordSender` were only needed because of a failure in our tests, not because of a real need. In tests we were injecting `TxnAbortedErrors` without the transaction record having actually been aborted. This meant that if a concurrent request attempted to heartbeat the txn record, it would find that it was still pending, which resulted in the deadlock mentioned previously.

I've changed the test to properly abort the meta record when it injects a `TxnAbortedError`, which allows us to remove the new hooks into `TxnCoordSender` and simplify the rest of this significantly. PTAL

tschottdorf commented:
LGTM except the last commit, for which you'll definitely want Andrei's eyes. Nice simplification though!

---

Reviewed 13 of 13 files at r5, 1 of 1 files at r6, 12 of 12 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-:-Ks9_orzNzWl338UK4Cj:b-5avz3k)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Reviewed 1 of 1 files at r6, 12 of 12 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 1096 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9viNc-YXCQ8WKqgSy:-Ks9viNc-YXCQ8WKqgSz:bjd9ulq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/internal/client/txn.go#L1096)):*
> ```Go
> 	newTxn := &retryErr.Transaction
> 
> 	abortedPreviouslyDuringTx := requestTxnID != txn.mu.Proto.ID
> ```

Get rid of the temporary variable; the comment does a better job of explaining what's  going on than the variable name does. 

---

*[pkg/roachpb/errors.proto, line 333 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9uGNkL7u4UZxpec_t:-Ks9uGNkL7u4UZxpec_u:b-dtiovo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L333)):*
> ```protobuf
> // transaction. This can happen in cases where concurrent requests are made
> // for a Transaction and one of the requests results in a Txn abort.
> message TxnAbortedAsyncErr {
> ```

`s/AbortedAsyncErr/PrevAttemptAbortedError/`? Or maybe just `TxnPrevAttemptError`, since the relationship between this condition and an abort is not obvious. 

(I think of regular TxnAbortedErrors as being asynchronous in some sense, because they're the result of an abort that is asynchronous with respect to the txn being aborted)

---

*[pkg/internal/client/sender.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWOEqUVp3ZUrCmkeen:-Ks9sTj9DzJJW7F5OQiC:bal6qbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L33)):*
> #17657

Close that issue if it's no longer relevant with the latest refactor of this change. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-:-Ks9xj08kQSej1MuMi-G:bb6ywmx)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 13 files at r5, 5 of 12 files at r7.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/errors.proto, line 273 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9nF8vOEXcl0lCjs0u:-Ks9nF8vOEXcl0lCjs0v:b-s0jb9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L273)):*
> ```protobuf
> }
> 
> // UnhandledRetryableError tells the recipient that a KV request must be
> ```

superfluous move of `UnhandledRetryableError` and `HandledRetryableError`? Nothing changed to them, right?
I'd try to move them back to avoid the diff.

---

*[pkg/roachpb/errors.proto, line 324 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9n_EjMHbZmkww0-7b:-Ks9n_EjMHbZmkww0-7c:bcmbcd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L324)):*
> ```protobuf
> // An UntrackedTxnError indicates that an attempt to transmit a batch request
> // for a Transaction failed because the Transaction was not being tracked by
> // the client properly.
> ```

consider clarifying "by the client" a bit - "by the TxnCoordSender"?

---

*[pkg/roachpb/errors.proto, line 363 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9nqwPhKPYulgyP1l_:-Ks9nqwPhKPYulgyP1la:by2jk3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L363)):*
> ```protobuf
>   optional ReplicaCorruptionError replica_corruption = 17;
>   optional ReplicaTooOldError replica_too_old = 18;
>   reserved 25;
> ```

consider putting these 3 lines back where they were

---

*[pkg/sql/executor.go, line 1366 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KsA-axjwYpgUr3_a5Br:-KsA-axjwYpgUr3_a5Bs:b-68ud6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/sql/executor.go#L1366)):*
> ```Go
> 			// the stmts and clear the parallel batch's error-set before
> 			// restarting.
> 			if parErr := session.synchronizeParallelStmts(session.context); parErr != nil {
> ```

add a TODO here to cancel the parallel queries?

---

*[pkg/sql/executor.go, line 1366 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KsA1xBu_mMKmK5n7SlA:-KsA1xBu_mMKmK5n7SlB:bt8m5gi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/sql/executor.go#L1366)):*
> ```Go
> 			// the stmts and clear the parallel batch's error-set before
> 			// restarting.
> 			if parErr := session.synchronizeParallelStmts(session.context); parErr != nil {
> ```

also comment that a big reason for the sync is that it'd be incorrect to run statements after the txn.Reset below cause the might write at the wrong epoch.

---

*[pkg/sql/executor.go, line 1405 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KsA2HgZ9gL87N0wUh1R:-KsA2HgZ9gL87N0wUh1S:bgt3xlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/sql/executor.go#L1405)):*
> ```Go
> 	_, independentFromParallelStmts := stmt.AST.(parser.IndependentFromParallelizedPriors)
> 	if !(parallelize || independentFromParallelStmts) {
> 		if err := session.synchronizeParallelStmts(session.context); err != nil {
> ```

I think you want `session.Ctx()` here and below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-:-KsA3_cahKNIRgtKtNam:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvb committed:
roachpb: group Unhandled/HandledRetryableError with other ErrorDetails

These two proto messages were not grouped where they should have been in
the file.

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 1096 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9viNc-YXCQ8WKqgSy:-KsFHlCSTYuBOy9m5QPJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/internal/client/txn.go#L1096)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Get rid of the temporary variable; the comment does a better job of explaining what's  going on than the variable name does. 
</blockquote></details>

Done.

---

*[pkg/roachpb/errors.proto, line 273 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9nF8vOEXcl0lCjs0u:-KsEjQinMqSS84G4Aao8:b14l50u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L273)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

superfluous move of `UnhandledRetryableError` and `HandledRetryableError`? Nothing changed to them, right?
I'd try to move them back to avoid the diff.
</blockquote></details>

I move them because the ordering was getting all messed up here. The pattern in this file is that the ErrorDeatil types were all listed up here in order. The mixed up order made it harder to read. Split into a separate commit for explicitness.

---

*[pkg/roachpb/errors.proto, line 324 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9n_EjMHbZmkww0-7b:-KsFFI8grukTCPaQpS1l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L324)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider clarifying "by the client" a bit - "by the TxnCoordSender"?
</blockquote></details>

Done.

---

*[pkg/roachpb/errors.proto, line 333 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9uGNkL7u4UZxpec_t:-KsFFX96YoMhJ-TmGaSL:b-pkelu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L333)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`s/AbortedAsyncErr/PrevAttemptAbortedError/`? Or maybe just `TxnPrevAttemptError`, since the relationship between this condition and an abort is not obvious. 

(I think of regular TxnAbortedErrors as being asynchronous in some sense, because they're the result of an abort that is asynchronous with respect to the txn being aborted)
</blockquote></details>

Done. Changes to `TxnPrevAttemptError`.

---

*[pkg/roachpb/errors.proto, line 363 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-Ks9nqwPhKPYulgyP1l_:-KsFFkvdfla7oJAZre5R:bhxsl13) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/roachpb/errors.proto#L363)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider putting these 3 lines back where they were
</blockquote></details>

These should all be in order, since this is just a giant union. The move happens in its own commit now though, for clarity of the blame.

---

*[pkg/sql/executor.go, line 1366 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KsA-axjwYpgUr3_a5Br:-KsFGh7-4suzexgP9ZQX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/sql/executor.go#L1366)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

add a TODO here to cancel the parallel queries?
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1366 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KsA1xBu_mMKmK5n7SlA:-KsFGxiPnX-OOwBoaFDo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/sql/executor.go#L1366)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

also comment that a big reason for the sync is that it'd be incorrect to run statements after the txn.Reset below cause the might write at the wrong epoch.
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1405 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KsA2HgZ9gL87N0wUh1R:-KsFHZmazt9QmvZMQHLF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6b6d99a55a6b0f0d3a6947e0398f8a3d6351fa/pkg/sql/executor.go#L1405)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think you want `session.Ctx()` here and below.
</blockquote></details>

Done.

---

*[pkg/internal/client/sender.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627#-KrWOEqUVp3ZUrCmkeen:-KsFHj2h1bO1AxM3i63j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/909cfc1ff87dc9bb135da04dec7769e87cd4d5b5/pkg/internal/client/sender.go#L33)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> #17657

Close that issue if it's no longer relevant with the latest refactor of this change. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17627)*
<!-- Sent from Reviewable.io -->

knz commented:
Isn't this that should be backported to 1.0?

nvanbenschoten commented:
@knz yes it should, along with #17868. What's the process for doing so?

knz commented:
1. Fork the branch `release-1.0`. 
2. Make an equivalent patch against that branch. (e.g. via `git cherry-pick`). 
3. Create a PR on github as usual; don't forget to select `release-1.0` as base branch for the PR). 
4. Prefix the PR title with "cherrypick". 
5. @cockroachdb/release in a comment to the PR.

nvanbenschoten commented:
👍 Thanks for the instructions!
